src/lib/support/BufferReader.h

```diff
@@ -0,0 +1,150 @@
+ */
+
+/**
+ *  @file
+ *    Utility classes for safely reading from size-limited buffers.
+ */
+
+#pragma once
+
+#include <climits>
+#include <lib/core/CHIPEncoding.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/CodeUtils.h>
+#include <stdint.h>
+
+namespace chip {
+namespace Encoding {
+namespace LittleEndian {
+
+/**
+ *  @class Reader
+ *
+ *  Simple reader for reading little-endian things out of buffers.
+ */
+class Reader
+{
+public:
+    /**
+     * Create a data model reader from a given buffer and length.
+     *
+     * @param buffer The octet buffer to read from.  The caller must ensure
+     *               (most simply by allocating the reader on the stack) that
+     *               the buffer outlives the reader.  The buffer is allowed to
+     *               be null if buf_len is 0.
+     * @param buf_len The number of octets in the buffer.
+     */
+    Reader(const uint8_t * buffer, uint16_t buf_len) : mBufStart(buffer), mReadPtr(buffer), mAvailable(buf_len) {}
+
+    /**
+     * Number of octets we have read so far.  This might be able to go away once
+     * we do less switching back and forth between DataModelReader and raw
+     * buffers.
+     */
+    uint16_t OctetsRead() const { return static_cast<uint16_t>(mReadPtr - mBufStart); }
+
+    /**
+     * Read a single 8-bit unsigned integer.
+     *
+     * @param [out] dest Where the 8-bit integer goes.
+     *
+     * @return Whether the read succeeded.  The read can fail if there are not
+     *         enough octets available.
+     */
+    CHECK_RETURN_VALUE
+    CHIP_ERROR Read8(uint8_t * dest) { return Read(dest); }
+
+    /**
+     * Read a single 16-bit unsigned integer.
+     *
+     * @param [out] dest Where the 16-bit integer goes.
+     *
+     * @return Whether the read succeeded.  The read can fail if there are not
+     *         enough octets available.
+     */
+    CHECK_RETURN_VALUE
+    CHIP_ERROR Read16(uint16_t * dest) { return Read(dest); }
+
+    /**
+     * Read a single 32-bit unsigned integer.
+     *
+     * @param [out] dest Where the 32-bit integer goes.
+     *
+     * @return Whether the read succeeded.  The read can fail if there are not
+     *         enough octets available.
+     */
+    CHECK_RETURN_VALUE
+    CHIP_ERROR Read32(uint32_t * dest) { return Read(dest); }
+
+    /**
+     * Read a single 64-bit unsigned integer.
+     *
+     * @param [out] dest Where the 64-bit integer goes.
+     *
+     * @return Whether the read succeeded.  The read can fail if there are not
+     *         enough octets available.
+     */
+    CHECK_RETURN_VALUE
+    CHIP_ERROR Read64(uint64_t * dest) { return Read(dest); }
+
+protected:
+    template <typename T>
+    CHIP_ERROR Read(T * retval)
+    {
+        static_assert(CHAR_BIT == 8, "Our various sizeof checks rely on bytes and octets being the same thing");
+
+        static constexpr size_t data_size = sizeof(T);
+
+        if (mAvailable < data_size)
```

Comment:
thought: should we somehow try to make errors persistent? If I have (wrong really) code like:

```cpp
Reader reader(buffer, 2);

reader.Read32(&foo);
return reader.Read16(&bar);
```

we do not notice that a value failed to be read in between. I imagine we generally want to either use some GetLength() call to figure out available length (if data is variable size) or once reader failed to read, it became invalid.

If we have persistent errors, we can have builder patters:

```cpp
Reader reader(buffer, len);

return reader.Read8(&foo).Read16(&bar).Read8(&baz).StatusCode();
```

instead of a more verbose VerifyOrExit:

```cpp
CHIP_ERROR err = CHIP_NO_ERROR;
Reader reader(buffer, len);

err = reader.Read8(&foo);
SuccessOrExit(err);

err = reader.Read16(&bar);
SuccessOrExit(err);

err = reader.Read8(&baz);
SuccessOrExit(err);

exit:
   return err;
```

I _could_have similar `SuccessOrExit(reader.StatusCode())` if I wanted to for some performance benefits, however I generally expect happy path to almost alwyas happens and would be willing to trade some ifs cost on failure for smaller flash usage.

Reply:
> If I have (wrong really) code like

That code won't compile, because `Read32` is marked as `CHECK_RETURN_VALUE` and in this code that return value is being ignored.

You could have code like this, though:
```
CHIP_RESULT res;
Reader reader(buffer, 2);

res = reader.Read32(&foo);
return reader.Read16(&bar);
```
but hopefully at that point you would notice that you are not doing anything with `res`.  And even better would be if we could make the compiler notice...

> If we have persistent errors, we can have builder patterns:

That does make sense, in cases when all the reads are unconditional.  In cases when the reads are conditional on the value a previous read returned, people would need to make sure to examine the success value of the previous read.  I guess we could maybe do that if we require that the reference we return will be used (for either an immediate unconditional read call or for an error check)?

One other note: I plan to build a data-model-specific reader on top of this one to address <https://github.com/project-chip/connectedhomeip/issues/2168>.  But I suspect that might also be possible with the reader pattern.  Will check.

I'll prototype out this suggestion and see how the code and the resulting codesize look.

Reply:
@andy31415 OK, I put up some example code at https://github.com/bzbarsky-apple/connectedhomeip/tree/little-endian-reader-builder and https://github.com/bzbarsky-apple/connectedhomeip/tree/data-model-message-reader-builder for the builder pattern approach.  It seems to use more codesize than the code I had before, though maybe it would do a bit better with some more out-of-lining, since `Read` ends up with a bit more logic.

Fundamentally, the builder setup still ends up doing the "did the previous call succeed" on every read check, which is what the `SuccessOrExit` was doing...

I did test instead setting `mAvailable` to 0 on read failure, so we don't have the extra branch on the status, but the end result is the same in terms of codesize; presumably the assignment is about the same size as the status-check branch.

Reply:
We can iterate. I would expect it to be smaller in size, but only if used a lot (so that the save of no extra if is worth it to extra logic inside reads). 

I am ok with the PR as is.

Reply:
OK, that works.  I'm pretty happy to revisit once we have more consumers and can get better size data.  Converting should not be that big a problem, I suspect....

Reply:
I did some more experimenting, and I think you're right and the builder pattern might be a better fit here.  Filed https://github.com/project-chip/connectedhomeip/issues/3495 and will put up a PR soon.